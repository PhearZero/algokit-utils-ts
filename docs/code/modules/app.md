[algotstest](../README.md) / app

# Module: app

## Table of contents

### Interfaces

- [AppCallArgs](../interfaces/app.AppCallArgs.md)
- [AppCallParams](../interfaces/app.AppCallParams.md)
- [AppReference](../interfaces/app.AppReference.md)
- [AppStorageSchema](../interfaces/app.AppStorageSchema.md)
- [BoxReference](../interfaces/app.BoxReference.md)
- [CompiledTeal](../interfaces/app.CompiledTeal.md)
- [CreateAppParams](../interfaces/app.CreateAppParams.md)
- [UpdateAppParams](../interfaces/app.UpdateAppParams.md)

### Variables

- [APP\_PAGE\_MAX\_SIZE](app.md#app_page_max_size)

### Functions

- [callApp](app.md#callapp)
- [compileTeal](app.md#compileteal)
- [createApp](app.md#createapp)
- [getAppArgsForTransaction](app.md#getappargsfortransaction)
- [getAppByIndex](app.md#getappbyindex)
- [updateApp](app.md#updateapp)

## Variables

### APP\_PAGE\_MAX\_SIZE

• `Const` **APP\_PAGE\_MAX\_SIZE**: ``2048``

#### Defined in

[app.ts:16](https://github.com/algorandfoundation/algokit-utils-ts/blob/4edaa90/src/app.ts#L16)

## Functions

### callApp

▸ **callApp**(`call`, `algod`): `Promise`<[`SendTransactionResult`](../interfaces/transaction.SendTransactionResult.md)\>

Issues a call to a given app.

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `call` | [`AppCallParams`](../interfaces/app.AppCallParams.md) | The call details. |
| `algod` | `default` | An algod client |

#### Returns

`Promise`<[`SendTransactionResult`](../interfaces/transaction.SendTransactionResult.md)\>

The result of the call

#### Defined in

[app.ts:204](https://github.com/algorandfoundation/algokit-utils-ts/blob/4edaa90/src/app.ts#L204)

___

### compileTeal

▸ **compileTeal**(`tealCode`, `algod`): `Promise`<[`CompiledTeal`](../interfaces/app.CompiledTeal.md)\>

Compiles the given TEAL using algod and returns the result.

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `tealCode` | `string` | The TEAL code |
| `algod` | `default` | An algod client |

#### Returns

`Promise`<[`CompiledTeal`](../interfaces/app.CompiledTeal.md)\>

The information about the compiled file

#### Defined in

[app.ts:276](https://github.com/algorandfoundation/algokit-utils-ts/blob/4edaa90/src/app.ts#L276)

___

### createApp

▸ **createApp**(`create`, `algod`): `Promise`<[`SendTransactionResult`](../interfaces/transaction.SendTransactionResult.md) & [`AppReference`](../interfaces/app.AppReference.md)\>

Creates a smart contract app, returns the details of the created app.

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `create` | [`CreateAppParams`](../interfaces/app.CreateAppParams.md) | The parameters to create the app with |
| `algod` | `default` | An algod client |

#### Returns

`Promise`<[`SendTransactionResult`](../interfaces/transaction.SendTransactionResult.md) & [`AppReference`](../interfaces/app.AppReference.md)\>

The details of the created app, or the transaction to create it if `skipSending`

#### Defined in

[app.ts:131](https://github.com/algorandfoundation/algokit-utils-ts/blob/4edaa90/src/app.ts#L131)

___

### getAppArgsForTransaction

▸ **getAppArgsForTransaction**(`args?`): `undefined` \| { `accounts`: `undefined` \| `string`[] = args.accounts; `appArgs`: `undefined` \| `Uint8Array`[] ; `boxes`: `undefined` \| `BoxReference`[] ; `foreignApps`: `undefined` \| `number`[] = args.apps; `foreignAssets`: `undefined` \| `number`[] = args.assets; `lease`: `undefined` \| `Uint8Array`  }

Returns the app args ready to load onto an app

**`See`**

object

#### Parameters

| Name | Type |
| :------ | :------ |
| `args?` | [`AppCallArgs`](../interfaces/app.AppCallArgs.md) |

#### Returns

`undefined` \| { `accounts`: `undefined` \| `string`[] = args.accounts; `appArgs`: `undefined` \| `Uint8Array`[] ; `boxes`: `undefined` \| `BoxReference`[] ; `foreignApps`: `undefined` \| `number`[] = args.apps; `foreignAssets`: `undefined` \| `number`[] = args.assets; `lease`: `undefined` \| `Uint8Array`  }

#### Defined in

[app.ts:239](https://github.com/algorandfoundation/algokit-utils-ts/blob/4edaa90/src/app.ts#L239)

___

### getAppByIndex

▸ **getAppByIndex**(`appIndex`, `algod`): `Promise`<[`ApplicationResponse`](../interfaces/algod_type.ApplicationResponse.md)\>

Gets the current data for the given app from algod.

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `appIndex` | `number` | The index of the app |
| `algod` | `default` | An algod client |

#### Returns

`Promise`<[`ApplicationResponse`](../interfaces/algod_type.ApplicationResponse.md)\>

The data about the app

#### Defined in

[app.ts:265](https://github.com/algorandfoundation/algokit-utils-ts/blob/4edaa90/src/app.ts#L265)

___

### updateApp

▸ **updateApp**(`update`, `algod`): `Promise`<[`SendTransactionResult`](../interfaces/transaction.SendTransactionResult.md)\>

Updates a smart contract app.

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `update` | [`UpdateAppParams`](../interfaces/app.UpdateAppParams.md) | The parameters to update the app with |
| `algod` | `default` | An algod client |

#### Returns

`Promise`<[`SendTransactionResult`](../interfaces/transaction.SendTransactionResult.md)\>

The transaction

#### Defined in

[app.ts:174](https://github.com/algorandfoundation/algokit-utils-ts/blob/4edaa90/src/app.ts#L174)
