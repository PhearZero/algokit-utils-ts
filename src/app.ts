import algosdk, {
  ABIArgument,
  ABIMethod,
  ABIMethodParams,
  Address,
  Algodv2,
  AtomicTransactionComposer,
  makeBasicAccountTransactionSigner,
  OnApplicationComplete,
  SuggestedParams,
  Transaction,
} from 'algosdk'
import { Buffer } from 'buffer'
import { AlgoKitConfig } from './config'
import {
  encodeTransactionNote,
  getSenderAddress,
  getTransactionParams,
  sendTransaction,
  SendTransactionFrom,
  SendTransactionParams,
  SendTransactionResult,
  TransactionNote,
} from './transaction'
import { ApplicationResponse } from './types/algod'

export const APP_PAGE_MAX_SIZE = 2048

/** Information about an Algorand app */
export interface AppReference {
  /** The index of the app */
  appIndex: number
  /** The Algorand address of the account associated with the app */
  appAddress: string
}

/**
 * A grouping of the app ID and name of the box in an Uint8Array
 */
export interface BoxReference {
  /**
   * A unique application index
   */
  appIndex: number
  /**
   * Name of box to reference
   */
  name: Uint8Array | string
}

/**
 * App call args with raw values (minus some processing like encoding strings as binary)
 */
export interface RawAppCallArgs {
  /** The address of any accounts to load in */
  accounts?: (string | Address)[]
  /** Any application arguments to pass through */
  appArgs?: (Uint8Array | string)[]
  /** Any box references to load */
  boxes?: BoxReference[]
  /** IDs of any apps to load into the foreignApps array */
  apps?: number[]
  /** IDs of any assets to load into the foreignAssets array */
  assets?: number[]
  /** The optional lease for the transaction */
  lease?: string | Uint8Array
}

/**
 * App call args for an ABI call
 */
export interface ABIAppCallArgs {
  /** The ABI method to call, either:
   *  * {method_name e.g. `hello`}; or
   *  * {method_signature e.g. `hello(string)string`} */
  method: ABIMethodParams | ABIMethod
  /** The ABI args to pass in */
  args: ABIArgument[]
  /** The optional lease for the transaction */
  lease?: string | Uint8Array
}

/** Arguments to pass to an app call either:
 *   * The raw app call values to pass through into the transaction (after processing); or
 *   * An ABI method definition (method and args)
 **/
export type AppCallArgs = RawAppCallArgs | ABIAppCallArgs

/** Base interface for common data passed to an app create or update. */
interface CreateOrUpdateAppParams extends SendTransactionParams {
  /** The account (with private key loaded) that will send the ÂµALGOs */
  from: SendTransactionFrom
  /** The approval program as raw teal (string) or compiled teal, base 64 encoded as a byte array (Uint8Array) */
  approvalProgram: Uint8Array | string
  /** The clear state program as raw teal (string) or compiled teal, base 64 encoded as a byte array (Uint8Array) */
  clearStateProgram: Uint8Array | string
  /** Optional transaction parameters */
  transactionParams?: SuggestedParams
  /** The (optional) transaction note */
  note?: TransactionNote
  /** The arguments passed in to the app call */
  args?: AppCallArgs
}

/** Parameters that are passed in when creating an app. */
export interface CreateAppParams extends CreateOrUpdateAppParams {
  /** The storage schema to request for the created app */
  schema: AppStorageSchema
}

/** Parameters that are passed in when updating an app. */
export interface UpdateAppParams extends CreateOrUpdateAppParams {
  /** The index of the app to update */
  appIndex: number
}

export interface AppCallParams extends SendTransactionParams {
  /** The index of the app to call */
  appIndex: number
  /** The type of call, everything except create (@see {createApp} ) and update (@see {updateApp} ) */
  callType: 'optin' | 'closeout' | 'clearstate' | 'delete' | 'normal'
  /** The account to make the call from */
  from: SendTransactionFrom
  /** Optional transaction parameters */
  transactionParams?: SuggestedParams
  /** The (optional) transaction note */
  note?: TransactionNote
  /** The arguments passed in to the app call */
  args?: AppCallArgs
}

/** Parameters representing the storage schema of an app. */
export interface AppStorageSchema {
  /** Restricts number of ints in per-user local state */
  localInts: number
  /** Restricts number of byte slices in per-user local state */
  localByteSlices: number
  /** Restricts number of ints in global state */
  globalInts: number
  /** Restricts number of byte slices in global state */
  globalByteSlices: number
  /** Any extra pages that are needed for the smart contract; if left blank then the right number of pages will be calculated based on the teal code size */
  extraPages?: number
}

/** Information about a compiled teal program */
export interface CompiledTeal {
  /** Original TEAL code */
  teal: string
  /** The compiled code */
  compiled: string
  /** The has returned by the compiler */
  compiledHash: string
  /** The base64 encoded code as a byte array */
  compiledBase64ToBytes: Uint8Array
}

/**
 * Creates a smart contract app, returns the details of the created app.
 * @param create The parameters to create the app with
 * @param algod An algod client
 * @returns The details of the created app, or the transaction to create it if `skipSending`
 */
export async function createApp(create: CreateAppParams, algod: Algodv2): Promise<SendTransactionResult & AppReference> {
  const { from, approvalProgram: approval, clearStateProgram: clear, schema, note, transactionParams, args, ...sendParams } = create

  const approvalProgram = typeof approval === 'string' ? (await compileTeal(approval, algod)).compiledBase64ToBytes : approval
  const clearProgram = typeof clear === 'string' ? (await compileTeal(clear, algod)).compiledBase64ToBytes : clear

  const transaction = algosdk.makeApplicationCreateTxnFromObject({
    approvalProgram: approvalProgram,
    clearProgram: clearProgram,
    numLocalInts: schema.localInts,
    numLocalByteSlices: schema.localByteSlices,
    numGlobalInts: schema.globalInts,
    numGlobalByteSlices: schema.globalByteSlices,
    extraPages: schema.extraPages ?? Math.floor((approvalProgram.length + clearProgram.length) / APP_PAGE_MAX_SIZE),
    onComplete: algosdk.OnApplicationComplete.NoOpOC,
    suggestedParams: await getTransactionParams(transactionParams, algod),
    from: getSenderAddress(from),
    note: encodeTransactionNote(note),
    ...getAppArgsForTransaction(args),
    rekeyTo: undefined,
  })

  const { confirmation } = await sendTransaction({ transaction, from, sendParams }, algod)
  if (confirmation) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const appIndex = confirmation['application-index']!

    AlgoKitConfig.getLogger(sendParams.suppressLog).debug(`Created app ${appIndex} from creator ${getSenderAddress(from)}`)

    return { transaction, confirmation, appIndex, appAddress: algosdk.getApplicationAddress(appIndex) }
  } else {
    return { transaction, appIndex: 0, appAddress: '' }
  }
}

/**
 * Updates a smart contract app.
 * @param update The parameters to update the app with
 * @param algod An algod client
 * @returns The transaction
 */
export async function updateApp(update: UpdateAppParams, algod: Algodv2): Promise<SendTransactionResult> {
  const { appIndex, from, approvalProgram: approval, clearStateProgram: clear, note, transactionParams, args, ...sendParams } = update

  const approvalProgram = typeof approval === 'string' ? (await compileTeal(approval, algod)).compiledBase64ToBytes : approval
  const clearProgram = typeof clear === 'string' ? (await compileTeal(clear, algod)).compiledBase64ToBytes : clear

  const transaction = algosdk.makeApplicationUpdateTxnFromObject({
    appIndex,
    approvalProgram: approvalProgram,
    clearProgram: clearProgram,
    suggestedParams: await getTransactionParams(transactionParams, algod),
    from: getSenderAddress(from),
    note: encodeTransactionNote(note),
    ...getAppArgsForTransaction(args),
    rekeyTo: undefined,
  })

  AlgoKitConfig.getLogger(sendParams.suppressLog).debug(`Updating app ${appIndex}`)

  return await sendTransaction({ transaction, from, sendParams }, algod)
}

/**
 * Issues a call to a given app.
 * @param call The call details.
 * @param algod An algod client
 * @returns The result of the call
 */
export async function callApp(call: AppCallParams, algod: Algodv2): Promise<SendTransactionResult> {
  const { appIndex, callType, from, args, note, transactionParams, ...sendParams } = call

  const appCallParameters = {
    appIndex: appIndex,
    from: getSenderAddress(from),
    suggestedParams: await getTransactionParams(transactionParams, algod),
    ...getAppArgsForTransaction(args),
    note: encodeTransactionNote(note),
    rekeyTo: undefined,
  }

  let transaction: Transaction
  switch (callType) {
    case 'optin':
      transaction = algosdk.makeApplicationOptInTxnFromObject(appCallParameters)
      break
    case 'clearstate':
      transaction = algosdk.makeApplicationClearStateTxnFromObject(appCallParameters)
      break
    case 'closeout':
      transaction = algosdk.makeApplicationCloseOutTxnFromObject(appCallParameters)
      break
    case 'delete':
      transaction = algosdk.makeApplicationDeleteTxnFromObject(appCallParameters)
      break
    case 'normal':
      transaction = algosdk.makeApplicationNoOpTxnFromObject(appCallParameters)
      break
  }

  return await sendTransaction({ transaction, from, sendParams }, algod)
}

/** Returns the app args ready to load onto an app @see {Transaction} object */
export function getAppArgsForTransaction(args?: AppCallArgs) {
  if (!args) return undefined

  let actualArgs: AppCallArgs
  if ('method' in args) {
    // todo: Land a change to algosdk that extract the logic from ATC, because (fair warning) this is a HACK
    // I don't want to have to rewrite all of the ABI resolution logic so using an ATC temporarily here
    // and passing stuff in to keep it happy like a randomly generated account :O
    // Most of these values aren't being used since the transaction is discarded
    const dummyAtc = new AtomicTransactionComposer()
    const dummyAccount = algosdk.generateAccount()
    const dummyAppId = 1
    const dummyParams = {
      fee: 1,
      firstRound: 1,
      genesisHash: Buffer.from('abcd', 'utf-8').toString('base64'),
      genesisID: 'a',
      lastRound: 1,
    }
    const dummyOnComplete = OnApplicationComplete.NoOpOC
    dummyAtc.addMethodCall({
      method: 'txnCount' in args.method ? args.method : new ABIMethod(args.method),
      methodArgs: args.args,
      // Rest are dummy values
      appID: dummyAppId,
      sender: dummyAccount.addr,
      signer: makeBasicAccountTransactionSigner(dummyAccount),
      suggestedParams: dummyParams,
      onComplete: dummyOnComplete,
    })
    const txn = dummyAtc.buildGroup()[0]
    actualArgs = {
      accounts: txn.txn.appAccounts,
      appArgs: txn.txn.appArgs,
      apps: txn.txn.appForeignApps,
      assets: txn.txn.appForeignAssets,
      boxes: txn.txn.boxes,
      lease: args.lease,
    }
  } else {
    actualArgs = args
  }

  const encoder = new TextEncoder()
  return {
    accounts: actualArgs?.accounts?.map((a) => (typeof a === 'string' ? a : algosdk.encodeAddress(a.publicKey))),
    appArgs: actualArgs?.appArgs?.map((a) => (typeof a === 'string' ? encoder.encode(a) : a)),
    boxes: actualArgs?.boxes?.map(
      (ref) =>
        ({
          appIndex: ref.appIndex,
          name: typeof ref.name === 'string' ? encoder.encode(ref.name) : ref.name,
        } as algosdk.BoxReference),
    ),
    foreignApps: actualArgs?.apps,
    foreignAssets: actualArgs?.assets,
    lease: typeof actualArgs?.lease === 'string' ? encoder.encode(actualArgs?.lease) : actualArgs?.lease,
  }
}

/**
 * Gets the current data for the given app from algod.
 *
 * @param appIndex The index of the app
 * @param algod An algod client
 * @returns The data about the app
 */
export async function getAppByIndex(appIndex: number, algod: Algodv2) {
  return (await algod.getApplicationByID(appIndex).do()) as ApplicationResponse
}

/**
 * Compiles the given TEAL using algod and returns the result.
 *
 * @param algod An algod client
 * @param tealCode The TEAL code
 * @returns The information about the compiled file
 */
export async function compileTeal(tealCode: string, algod: Algodv2): Promise<CompiledTeal> {
  const compiled = await algod.compile(tealCode).do()
  return {
    teal: tealCode,
    compiled: compiled.result,
    compiledHash: compiled.hash,
    compiledBase64ToBytes: new Uint8Array(Buffer.from(compiled.result, 'base64')),
  }
}
