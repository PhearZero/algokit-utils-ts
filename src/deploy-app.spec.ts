import { describe, test } from '@jest/globals'
import { getBareCallContractCreateParams, getBareCallContractData } from '../tests/example-contracts/bare-call/contract'
import { localNetFixture } from '../tests/fixtures/localnet-fixture'
import { callApp, createApp, updateApp } from './app'
import { deployApp, getCreatorAppsByName } from './deploy-app'

describe('deploy-app', () => {
  const localnet = localNetFixture()

  const name = 'MY_APP'

  test('Created app is retrieved by name with deployment metadata', async () => {
    const { client, indexer, testAccount, waitForIndexer } = localnet.context
    const creationMetadata = { name, version: '1.0', updatable: true, deletable: false }
    const app1 = await createApp(await getBareCallContractCreateParams(testAccount, creationMetadata), client)
    await waitForIndexer()

    const apps = await getCreatorAppsByName(indexer, testAccount)

    expect(apps.creator).toBe(testAccount.addr)
    expect(Object.keys(apps.apps)).toEqual([name])
    const app = apps.apps[name]
    expect(app.appIndex).toBe(app1.appIndex)
    expect(app.appAddress).toBe(app1.appAddress)
    expect(app.createdRound).toBe(app1.confirmation?.['confirmed-round'])
    expect(app.createdMetadata).toEqual(creationMetadata)
    expect(app.updatedRound).toBe(app.createdRound)
    expect(app.name).toBe(creationMetadata.name)
    expect(app.updatable).toBe(creationMetadata.updatable)
    expect(app.deletable).toBe(creationMetadata.deletable)
    expect(app.version).toBe(creationMetadata.version)
  })

  test('Created, updated and deleted apps are retrieved by name with deployment metadata', async () => {
    const { client, indexer, testAccount, waitForIndexer } = localnet.context
    const creationMetadata = { name, version: '1.0', updatable: true, deletable: true }
    const name2 = 'APP_2'
    const name3 = 'APP_3'
    const app1 = await createApp(await getBareCallContractCreateParams(testAccount, creationMetadata), client)
    const app2 = await createApp(await getBareCallContractCreateParams(testAccount, { ...creationMetadata, name: name2 }), client)
    const app3 = await createApp(await getBareCallContractCreateParams(testAccount, { ...creationMetadata, name: name3 }), client)
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const updateMetadata = { name, version: '2.0', updatable: false, deletable: false }
    const update1 = await updateApp(
      { ...(await getBareCallContractCreateParams(testAccount, updateMetadata)), appIndex: app1.appIndex },
      client,
    )
    const delete3 = await callApp({ appIndex: app3.appIndex, callType: 'delete', from: testAccount }, client)
    await waitForIndexer()

    const apps = await getCreatorAppsByName(indexer, testAccount)

    expect(apps.creator).toBe(testAccount.addr)
    expect(Object.keys(apps.apps).sort()).toEqual([name, name2, name3].sort())
    const app1Data = apps.apps[name]
    expect(app1Data.appIndex).toBe(app1.appIndex)
    expect(app1Data.appAddress).toBe(app1.appAddress)
    expect(app1Data.createdRound).toBe(app1.confirmation?.['confirmed-round'])
    expect(app1Data.createdMetadata).toEqual(creationMetadata)
    expect(app1Data.updatedRound).toBe(update1.confirmation?.['confirmed-round'])
    expect(app1Data.name).toBe(updateMetadata.name)
    expect(app1Data.updatable).toBe(updateMetadata.updatable)
    expect(app1Data.deletable).toBe(updateMetadata.deletable)
    expect(app1Data.version).toBe(updateMetadata.version)
    expect(app1Data.deleted).toBe(false)

    const app2Data = apps.apps[name2]
    expect(app2Data.appIndex).toBe(app2.appIndex)
    expect(app2Data.deleted).toBe(false)

    const app3Data = apps.apps[name3]
    expect(app3Data.appIndex).toBe(app3.appIndex)
    expect(app3Data.deleted).toBe(true)
  })

  test('Deploy new app', async () => {
    const { client, indexer, testAccount, waitForIndexer } = localnet.context
    const contract = await getBareCallContractData()

    const result = await deployApp(
      {
        approvalProgram: contract.approvalProgram,
        clearStateProgram: contract.clearStateProgram,
        from: testAccount,
        metadata: {
          name: 'test',
          version: '1.0',
          deletable: false,
          updatable: false,
        },
        schema: contract.stateSchema,
        deployTimeParameters: {
          VALUE: 1,
        },
      },
      client,
      indexer,
    )
  })
})
