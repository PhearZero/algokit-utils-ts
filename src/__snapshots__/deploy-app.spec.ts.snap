// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deploy-app Deploy delete to deletable updated app 1`] = `
"INFO: Idempotently deploying app "test" from creator ACCOUNT_1 using 1039 bytes of teal code and 39 bytes of teal code
INFO: Existing app test found by creator ACCOUNT_1, with app index APP_1 and version 1.0.
INFO: Detected a TEAL update in app APP_1 for creator ACCOUNT_1
WARN: App is deletable and onUpdate=DeleteApp, creating new app and deleting old app...
INFO: Deploying a new test app for ACCOUNT_1; deploying app with version 2.0.
WARN: Deleting existing test app with index APP_1 from ACCOUNT_1 account.
WARN: Sent transactions TXID_2 to create app with index APP_2 and TXID_3 to delete app with index APP_1 from ACCOUNT_1 account."
`;

exports[`deploy-app Deploy failure for app delete fails if onSchemaBreak = Fail 1`] = `
"INFO: Idempotently deploying app "test" from creator ACCOUNT_1 using 1039 bytes of teal code and 39 bytes of teal code
INFO: Existing app test found by creator ACCOUNT_1, with app index APP_1 and version 1.0.
WARN: Detected a breaking app schema change in app APP_1: | [{"from":{"global":{"num-byte-slice":0,"num-uint":1},"local":{"num-byte-slice":0,"num-uint":0}},"to":{"global":{"num-byte-slice":1,"num-uint":1},"local":{"num-byte-slice":0,"num-uint":0}}}]"
`;

exports[`deploy-app Deploy failure for delete to permanent app 1`] = `
"INFO: Idempotently deploying app "test" from creator ACCOUNT_1 using 1039 bytes of teal code and 39 bytes of teal code
INFO: Existing app test found by creator ACCOUNT_1, with app index APP_1 and version 1.0.
INFO: Detected a TEAL update in app APP_1 for creator ACCOUNT_1
WARN: App is not deletable and onUpdate=DeleteApp, will attempt to create new app and delete old app, delete will most likely fail
INFO: Deploying a new test app for ACCOUNT_1; deploying app with version 2.0.
WARN: Deleting existing test app with index APP_1 from ACCOUNT_1 account."
`;

exports[`deploy-app Deploy failure for updated app fails if onupdate = Fail 1`] = `
"INFO: Idempotently deploying app "test" from creator ACCOUNT_1 using 1039 bytes of teal code and 39 bytes of teal code
INFO: Existing app test found by creator ACCOUNT_1, with app index APP_1 and version 1.0.
INFO: Detected a TEAL update in app APP_1 for creator ACCOUNT_1"
`;

exports[`deploy-app Deploy new app 1`] = `
"INFO: Idempotently deploying app "test" from creator ACCOUNT_1 using 1039 bytes of teal code and 39 bytes of teal code
INFO: App test not found in apps created by ACCOUNT_1; deploying app with version 1.0.
INFO: Sent transaction ID TXID_1 appl from ACCOUNT_1
DEBUG: Created app APP_1 from creator ACCOUNT_1"
`;

exports[`deploy-app Deploy update to immutable updated app fails 1`] = `
"INFO: Idempotently deploying app "test" from creator ACCOUNT_1 using 1039 bytes of teal code and 39 bytes of teal code
INFO: Existing app test found by creator ACCOUNT_1, with app index APP_1 and version 1.0.
INFO: Detected a TEAL update in app APP_1 for creator ACCOUNT_1
WARN: App is not updatable but onUpdate=UpdateApp, will attempt to update app, update will most likely fail
INFO: Updating existing test app for ACCOUNT_1 to version 2.0.
DEBUG: Updating app APP_1"
`;

exports[`deploy-app Deploy update to updatable updated app 1`] = `
"INFO: Idempotently deploying app "test" from creator ACCOUNT_1 using 1039 bytes of teal code and 39 bytes of teal code
INFO: Existing app test found by creator ACCOUNT_1, with app index APP_1 and version 1.0.
INFO: Detected a TEAL update in app APP_1 for creator ACCOUNT_1
INFO: App is updatable and onUpdate=UpdateApp, updating app...
INFO: Updating existing test app for ACCOUNT_1 to version 2.0.
DEBUG: Updating app APP_1
INFO: Sent transaction ID TXID_2 appl from ACCOUNT_1"
`;
